server/routes/forms/forms.ts

```diff
@@ -1,26 +1,91 @@
-export default abstract class Form<Data extends { formId: string }> {
-  // posted form data; NB: it may be transformed in place by validation
-  data: Partial<Data>
+export interface BaseFormData {
+  /**
+   * Unique identifier to distinguish multiple forms on one page
+   */
+  formId: string
+}
 
-  // fields with error messages; for use with error message components
-  fieldErrors: Partial<Record<keyof Data, string>>
+/**
+ * Base form providing simple validation extension points and per-field error messages
+ */
+export default abstract class Form<Data extends BaseFormData> {
+  /**
+   * Posted form data, undefined indicates that the form has not been submitted/POSTed
+   * NB: it may be transformed in place by validation
+   */
+  protected data?: Partial<Data>
 
-  constructor(data: Partial<Data>) {
-    this.data = data
+  /**
+   * Holds per-field error messages
+   */
+  private readonly fieldErrors: Partial<Record<keyof Data, string>>
+
+  constructor(
+    /**
+     * Unique identifier to distinguish multiple forms on one page
+     */
+    readonly formId: string
+  ) {
     this.fieldErrors = {}
+  }
+
+  public toString(): string {
+    return `[Form formId=${this.formId} submitted=${this.submitted} hasErrors=${this.hasErrors}]`
```

Comment:
Maybe I'm missing some JS subtlety here but shouldn't `this.submitted()`/`this.hasErrors()` be invoked, e.g. have `()` after their names?

I don't imagine the string interpolation would do that automatically 🤔 ?

Reply:
they're both getter methods, so `form.submitted` calls the method

Reply:
Cheers, I missed that! Nice ☺️